parquet-column/src/main/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,484 @@
+  }
+
+  /**
+   * Construct the bloom filter with given bit set.
+   * @param input The given bitset to construct bloom filter.
+   * @param hash The hash strategy bloom filter apply.
+   * @param algorithm The algorithm of bloom filter.
+   */
+  public Bloom(ByteBuffer input, HASH hash, ALGORITHM algorithm) {
+    if (input.isDirect()){
+      bitset = new byte[input.remaining()];
+      input.get(bitset);
+    } else {
+      this.bitset = input.array();
+    }
+    this.numBytes = input.array().length;
+    this.bloomFilterHash = hash;
+    this.bloomFilterAlgorithm = algorithm;
+  }
+
+  /**
+   * Create a new bitset for bloom filter, at least 256 bits will be create.
+   * @param numBytes number of bytes for bit set.
+   */
+  public void initBitset(int numBytes) {
+    if (numBytes < MINIMUM_BLOOM_SIZE) {
+      numBytes = MINIMUM_BLOOM_SIZE;
+    }
+
+    if (numBytes > ParquetProperties.DEFAULT_MAXIMUM_BLOOM_FILTER_SIZE) {
+      numBytes = ParquetProperties.DEFAULT_MAXIMUM_BLOOM_FILTER_SIZE;
+    }
+
+    // 32 bytes alignment, one bucket.
+    numBytes = (numBytes + BYTES_PER_BUCKET - 1) & ~(BYTES_PER_BUCKET - 1);
+
+    ByteBuffer bytes = ByteBuffer.allocate(numBytes);
+    this.bitset = bytes.array();
+    this.numBytes = numBytes;
+  }
+
+  /**
+   * Create bitset from a given byte array.
+   * @param bitset Given bitset for bloom filter.
+   */
+  public void initBitset(byte[] bitset) {
+    this.bitset = bitset;
+    this.numBytes = bitset.length;
+  }
+
+  /**
+   * @return Bytes in buffered in this bloom filter.
+   */
+  public BytesInput getBytes() {
+    inputs.clear();
+
+    // Add number of bytes.
+    inputs.add(BytesInput.fromInt(numBytes));
+
+    // Add hash strategy
+    inputs.add(BytesInput.fromInt(this.bloomFilterHash.ordinal()));
+
+    // Add bloom filter algorithm
+    inputs.add(BytesInput.fromInt(this.bloomFilterAlgorithm.ordinal()));
+
+    // Add bitset.
+    inputs.add(BytesInput.from(bitset, 0, bitset.length));
+
+    return BytesInput.concat(inputs);
+  }
+
+  private void setMask(int key) {
+    final int SALT[] = {0x47b6137b, 0x44974d91, 0x8824ad5b, 0xa2b7289d,
+      0x705495c7, 0x2df1424b, 0x9efc4947, 0x5c6bfb31};
+
+    Arrays.fill(mask, 0);
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = key * SALT[i];
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = mask[i] >> 27;
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = 0x1 << mask[i];
+    }
+
+  }
+
+  /**
+   * Add an element to bloom filter, the element content is represented by
+   * the hash value of its plain encoding result.
+   * @param hash hash result of element.
+   */
+  public void bloomInsert(long hash) {
+    int bucketIndex = (int)(hash >> 32) & (numBytes / BYTES_PER_BUCKET - 1);
+    int key = (int)hash;
+
+    setMask(key);
+
+    for (int i = 0; i < 8; i++) {
+      int bitsetIndex = bucketIndex * BYTES_PER_BUCKET + i * 4;
+      bitset[bitsetIndex] |= (byte)(mask[i] >>> 24);
```

Comment:
Java use big endian so we need to handle the byte order inside integer. Otherwise we need to allocate temporary variable to store reverted order bytes.

Reply:
https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html#reverseBytes-int-

Reply:
I checked ByteBuffer APIs again, we can skip reverseBytes from Integer since ByteBuffer wrap bytes in BIG_ENDIAN order. But the bytes streaming OR is faster than integer reverse and OR in unit test.  testBinaryBloom takes average 200ms  vs 168ms.

Following is int OR: 
int origin = ByteBuffer.wrap(bitset, bitsetIndex, 4).asIntBuffer().get();
origin |= mask[i];
ByteBuffer.wrap(bitset, bitsetIndex, 4).asIntBuffer().put(origin);

This may because of the function call overhead.

Reply:
What if it is an int array rather than a byte buffer? That might more naturally match the semantics of the tiny split Bloom filter.

Reply:
Yes, it looks more semantics for the 32 bytes block. 
But it needs to convert endianness when serializing and deserializing which may need extra buffer.